gobblin-yarn/src/main/java/gobblin/yarn/GobblinWorkUnitRunner.java

```diff
@@ -0,0 +1,413 @@
+
+  private final ContainerId containerId;
+
+  private final HelixManager helixManager;
+
+  private final ServiceManager serviceManager;
+
+  private final EventBus eventBus;
+
+  private final TaskStateModelFactory taskStateModelFactory;
+
+  private final MetricRegistry metricRegistry = new MetricRegistry();
+  private final JmxReporter jmxReporter = JmxReporter.forRegistry(this.metricRegistry)
+      .convertRatesTo(TimeUnit.SECONDS)
+      .convertDurationsTo(TimeUnit.MILLISECONDS)
+      .build();
+
+  private volatile boolean stopInProgress = false;
+  private volatile boolean isStopped = false;
+
+  public GobblinWorkUnitRunner(String applicationName, Config config) throws Exception {
+    this.containerId =
+        ConverterUtils.toContainerId(System.getenv().get(ApplicationConstants.Environment.CONTAINER_ID.key()));
+    FileSystem fs = config.hasPath(ConfigurationKeys.FS_URI_KEY) ?
+        FileSystem.get(URI.create(config.getString(ConfigurationKeys.FS_URI_KEY)), new Configuration()) :
+        FileSystem.get(new Configuration());
+
+    String zkConnectionString = config.getString(ConfigurationConstants.ZK_CONNECTION_STRING_KEY);
+    LOGGER.info("Using ZooKeeper connection string: " + zkConnectionString);
+
+    this.helixManager = HelixManagerFactory
+        .getZKHelixManager(config.getString(ConfigurationConstants.HELIX_CLUSTER_NAME_KEY),
+            YarnHelixUtils.getParticipantIdStr(YarnHelixUtils.getHostname(), this.containerId),
+            InstanceType.PARTICIPANT, zkConnectionString);
+
+    Properties properties = YarnHelixUtils.configToProperties(config);
+
+    this.eventBus = new EventBus();
+
+    TaskExecutor taskExecutor = new TaskExecutor(properties);
+    TaskStateTracker taskStateTracker = new GobblinHelixTaskStateTracker(properties, this.helixManager);
+
+    List<Service> services = Lists.newArrayList();
+    if (config.hasPath(ConfigurationConstants.TOKEN_FILE_PATH)) {
+      LOGGER.info("Adding YarnContainerSecurityManager since login is keytab based");
+      services.add(new YarnContainerSecurityManager(config, fs, this.eventBus));
+    }
+    services.add(taskExecutor);
+    services.add(taskStateTracker);
+    this.serviceManager = new ServiceManager(services);
+
+    // Register task factory for the Helix task state model
+    Map<String, TaskFactory> taskFactoryMap = Maps.newHashMap();
+    Path appWorkDir = YarnHelixUtils.getAppWorkDirPath(
+        fs, applicationName, containerId.getApplicationAttemptId().getApplicationId());
+    taskFactoryMap.put(GOBBLIN_TASK_FACTORY_NAME,
+        new GobblinHelixTaskFactory(taskExecutor, taskStateTracker, fs, appWorkDir));
+    this.taskStateModelFactory = new TaskStateModelFactory(this.helixManager, taskFactoryMap);
+    this.helixManager.getStateMachineEngine().registerStateModelFactory(StateModelDefId.from("Task"),
+        this.taskStateModelFactory);
+  }
+
+  /**
+   * Start this {@link GobblinWorkUnitRunner} instance.
+   */
+  public void start() {
```

Comment:
This looks fairly similar to `GobblinApplicationMaster.start()`. Can this be refactored?